stackrox/stackrox
# ROX-24860: Add NVD CVSS column to Workload CVEs

### Description

Reduced scope to corresponding backend changes in #12770

### In scope

1. Although **Node CVEs** is out of scope, replace inconsistent **CVSS score** with **CVSS** column heading.
2. Add frontend business logic (mostly via copy-paste-edit).
    Besides replacement of property names, the biggest change is special case that `scoreVersion === 'UNKNOWN_VERSION'` means NVD CVSS scope is not available.
    * `CVSSFormatted` element encapsulates visible **Not available**.
    * `getScoreVersionsForTopNvdCVSS` function for `CVEsTable` element.
    * `getHighestNvdCvssScore` function for `AffectedImagesTable` element.
3. Move columns in one table so **CVE status** follows **CVE severity** like everywhere else.

### Out of scope

Workload CVEs
* /main/vulnerabilities/workload-cves?entityTab=Deployment
    POST /api/graphql?opname=getCvesForDeployment `nvdCvss`

Node CVEs
* /main/vulnerabilities/node-cves?entityTab=CVE
     POST /api/graphql?opname=getNodeCVEs `topNvdCvss`
* In table body, click CVE link
    POST /api/graphql?opname=getAffectedNodes `nvdCvss`
* In table body, click node link.
    POST /api/graphql?opname=getNodeVulnerabilities `nvdCvss`

### Testing

### User-facing documentation


### Testing and quality


#### Automated testing


#### How I validated my change

Prerequisite before `npm run deploy` in ui folder

```sh
export ROX_NVD_CVSS_UI=true
```

1. `npm run lint` in ui/apps/platform
5. `npm run build` in ui/apps/platform
6. `npm run start` in ui/apps/platform

#### Manual testing

1. Visit /main/vulnerabilities/workload-cves?entityTab=CVE see `CVEsTable` element.
    POST /api/graphql?opname=getImageCVEList response has `topCVSS` property only.

    * Before changes, see absence of **Top NVD CVSS** column.
        ![WorkloadCVEs_CVEsTable_without_Top_NVD_CVSS](https://github.com/user-attachments/assets/edd4b5c7-77c5-4718-bf39-2f5286629a68)

    * After changes, see presence **Top NVD CVSS** column with placeholder **Not available** value.
        ![WorkloadCVEs_CVEsTable_with_Top_NVD_CVSS](https://github.com/user-attachments/assets/6573d844-28f0-4ee2-a928-484e76456304)

2. In table body, click CVE link: see `AffectedImagesTable` element.
    POST /api/graphql?opname=getImagesForCVE response has `cvss` property only.
    Compare **CVE status** column to next step.

    * Before changes, see absence of **NVD CVSS** column and **CVE status** separated by score from **CVE severity**.
        ![AffectedImagesTable_without_NVD_CVSS](https://github.com/user-attachments/assets/d858601b-0e15-4e4d-9f18-b7b63e3ee47b)

    * After changes, see presence of **NVD CVSS** column with placeholder **Not available** value and **CVE status** follows **CVE severity**.
        ![AffectedImagesTable_with_NVD_CVSS](https://github.com/user-attachments/assets/5db026b0-f46b-4f9d-b903-b5302c95f0d7)

3. In table body, click image link: see `ImageVulnerabilitiesTable` element.
    POST /api/graphql?opname=getCVEsForImage response has `cvss` property only.
    See **CVE status** follows **CVE severity**.

    * Before changes, see absence of **NVD CVSS** column.
        ![ImageVulnerabilitiesTable_without_NVD_CVSS](https://github.com/user-attachments/assets/8510cfd7-662e-464f-8a88-94705a53bfc7)

    * After changes, see presence of **NVD CVSS** column with placeholder **Not available** value.
        ![ImageVulnerabilitiesTable_with_NVD_CVSS](https://github.com/user-attachments/assets/8d7672ef-4d1d-44a4-892c-fbfaa28a3dfb)

#### Integration testing

Prerequisite to `npm run cypress-open` in ui/apps/platform folder:

```sh
export CYPRESS_ROX_VULN_MGMT_ADVANCED_FILTERS=true
export CYPRESS_ROX_VULN_MGMT_NODE_PLATFORM_CVES=true
export CYPRESS_ROX_VULN_MGMT_UNIFIED_CVE_DEFERRAL=true
export CYPRESS_ROX_VULN_MGMT_WORKLOAD_CVES=true
export CYPRESS_ROX_WORKLOAD_CVES_FIXABILITY_FILTERS=true
```

* vulnerabilities/nodeCves/cveDetailPage.test.ts
* vulnerabilities/nodeCves/nodeCvesOverviewPage.test.ts
* vulnerabilities/nodeCves/nodeDetailPage.test.ts
* vulnerabilities/workloadCves/cveExceptionFlow.test.js
* vulnerabilities/workloadCves/deploymentSingle.test.ts
* vulnerabilities/workloadCves/imageCveListExceptionFlow.test.ts
* vulnerabilities/workloadCves/imageCveSingle.test.ts
* vulnerabilities/workloadCves/imageSingle.test.ts
* vulnerabilities/workloadCves/namespaceView.test.ts
* vulnerabilities/workloadCves/watchedImageFlow.test.ts
* vulnerabilities/workloadCves/workloadCveOverviewPage.test.ts

#### Component testing

Thank you, **David** for directing my attention to example of feature flag.

* src/Containers/Vulnerabilities/NodeCves/NodeCve/AffectedNodesTable.cy.jsx
* src/Containers/Vulnerabilities/NodeCves/components/NodeComponentsTable.cy.jsx
* src/Containers/Vulnerabilities/WorkloadCves/components/WorkloadCveFilterToolbar.cy.jsx
* src/Containers/Vulnerabilities/components/CvePageHeader.cy.jsx
* src/Containers/Vulnerabilities/components/FilterAutocomplete.cy.js
* src/Containers/Vulnerabilities/components/SnoozedCveToggleButton.cy.jsx
* src/Containers/Vulnerabilities/components/SummaryCardLayout.cy.jsx
